### PR TITLE
Fix early GC collection of Reader

### PR DESCRIPTION
**Pull request by erikcorry**
Adds a refcounted reference to the Reader object to stop it going away too early.

This bug was detected with a gc-stress mode on a test in streams-test.ts:
"ReadableStream start should be able to return a promise and reject it"

The test is as follows:

```js
promise_test(() => {
  const theError = new Error('rejected!');
  const rs = new ReadableStream({
    start() {
      return delay(1).then(() => { throw theError; });
    }
  });
  return rs.getReader().closed.then(() => {
    assert_unreached('closed promise should be rejected');
  }, e => {
    assert_equals(e, theError, 'promise should be rejected with the same error');
  });
}, 'ReadableStream start should be able to return a promise and reject it');
```

The root cause is that the Reader has a reference to the Stream, but the Stream has no reference to the Reader. If the user does not keep a reference to the Reader, it can be collected early, which causes a destructor to remove promises from the queue and never resolve them.

When the Reader is GC'd, the following cascade occurs:

1. V8 GC collects the Reader's JS wrapper (unreachable from JS roots)
2. CppGC collects the shim → `detachWrapper()` on the Reader's Wrappable
3. Reader C++ object is destroyed:
   - `ReaderImpl::~ReaderImpl()` destroys `closedPromise` (the C++ handle; the JS promise object survives, held by the `.then()` chain)
   - `ReaderImpl::~ReaderImpl()` destroys `state`, which is `Attached`
   - `Attached::~Attached()` destroys `jsg::Ref<ReadableStream>` — the **last** ref to the stream (the `rs` local variable went out of scope when the test function returned)
4. `ReadableStream` is destroyed:
   - `ReadableStreamJsController` (member) is destroyed
   - `kj::Own<ValueReadable>` state is destroyed
   - `ValueReadable::~ValueReadable()` destroys its `Consumer`
   - `~ConsumerImpl()` calls `queue.removeConsumer(*this)` (queue.h:390) — **removes itself from the Controller's queue**
   - `ReadableLockImpl` (the `lock` member) is destroyed
   - `ReaderLocked::~ReaderLocked()` calls `reader.detach()` — but the Reader is already destroyed, so this is a **use-after-free** on the raw `Reader&` (common.h:815). In practice this likely doesn't crash because the memory hasn't been reused yet.
   - `ReaderLocked::~ReaderLocked()` then implicitly destroys `closedFulfiller` (the Resolver) **without ever resolving or rejecting it**

5. Later: `delay(1)` fires → `start()` promise rejects → `onFailure` lambda fires → `ReadableImpl::doError()` on the Controller (which is alive, held by the lambda's `Ref<ReadableStreamDefaultController>`) → `queue.error()` → iterates consumers → **no consumers left** (removed in step 4) → `onConsumerError` is never called → `ReadableStreamJsController::doError` is never called → `lock.onError()` is never called → `closedFulfiller` was already destroyed

### CODE REVIEW

**Comment by erikcorry** [comment]
Looks like this causes a leak, caught by the internal build. Trying to see how to fix it.

**Comment by erikcorry** [comment]
Needs a bigger refactor
